### PR TITLE
Fix #3947: Support lifting of java.lang.Class

### LINKED ISSUE #3947: Lift a Class[T] to an Expr[Class[T]]
We need to be able to lift class tags to be able to lift `Array`.  A `ClassTag` can be constructed with a `java.lang.Class`, hence we should be able to lift a `java.lang.Class`.

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Missing a test case for lifted anonymous classes. These have an unrelated bug #4192.

File: compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala [hunk 1]
```diff
@@ -33,7 +34,11 @@ object PickledQuotes {
   /** Transform the expression into its fully spliced Tree */
   def quotedExprToTree(expr: quoted.Expr[_])(implicit ctx: Context): Tree = expr match {
     case expr: TastyExpr[_] => unpickleExpr(expr)
-    case expr: LiftedExpr[_] => Literal(Constant(expr.value))
+    case expr: LiftedExpr[_] =>
+      expr.value match {
+        case value: Class[_] => ref(defn.Predef_classOf).appliedToType(classToType(value))
+        case value => Literal(Constant(value))
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala]: Did you know that you can pass a `Type` to the constructor of `Constant` and if you do `Constant#tpe` on such a value you get an instance of `java.lang.Class` ?

Comment (smarter) [reply compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala]: The miniphase `ClassOf` rewrites call to `classOf` into `Literal(Constant(...))`

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala]: I will look at those.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala]: I can use the representation with `Literal(Constant(...))` from `ClassOf` though this representation is only expected after this phase. It seems better to create the `classOf[T]` and let `ClassOf` rewrite all of them at that phase.

File: compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala [hunk 2]
```diff
@@ -154,4 +159,25 @@ object PickledQuotes {
...
+    if (clazz == classOf[Boolean]) defn.BooleanType
+    else if (clazz == classOf[Byte]) defn.ByteType
+    else if (clazz == classOf[Char]) defn.CharType
+    else if (clazz == classOf[Short]) defn.ShortType
+    else if (clazz == classOf[Int]) defn.IntType
+    else if (clazz == classOf[Long]) defn.LongType
+    else if (clazz == classOf[Float]) defn.FloatType
+    else if (clazz == classOf[Double]) defn.DoubleType
+    else if (clazz == classOf[Unit]) defn.UnitType
+    else if (!clazz.isMemberClass) ctx.getClassIfDefined(clazz.getCanonicalName).typeRef
```
Comment (retronym) [inline compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala]: `Array`-s probably need special handling, too.

File: compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala [hunk 3]
```diff
@@ -154,4 +159,25 @@ object PickledQuotes {
...
+  private def classToType(clazz: Class[_])(implicit ctx: Context): Type = {
+    if (clazz == classOf[Boolean]) defn.BooleanType
+    else if (clazz == classOf[Byte]) defn.ByteType
+    else if (clazz == classOf[Char]) defn.CharType
+    else if (clazz == classOf[Short]) defn.ShortType
+    else if (clazz == classOf[Int]) defn.IntType
+    else if (clazz == classOf[Long]) defn.LongType
+    else if (clazz == classOf[Float]) defn.FloatType
+    else if (clazz == classOf[Double]) defn.DoubleType
+    else if (clazz == classOf[Unit]) defn.UnitType
```
Comment (retronym) [inline compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala]: Could enclose the above in `if (clazz.isPrimitive)` for performance reasons.

File: compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala (hunk 2 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala]: Will look into it

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala]: Yes, they needed it. Thanks :)

File: tests/run-with-compiler/i3947.scala [hunk 4]
```diff
@@ -0,0 +1,24 @@
+
+import scala.quoted._
+import dotty.tools.dotc.quoted.Toolbox._
+
+object Test {
+
+  def main(args: Array[String]): Unit = {
+
+    def test[T](clazz: java.lang.Class[T]): Unit = {
+      val lclazz = clazz.toExpr
```
Comment (biboudis) [inline tests/run-with-compiler/i3947.scala]: What's the purpose of `lclazz` in the test?

File: tests/run-with-compiler/i3947b.scala [hunk 5]
```diff
@@ -0,0 +1,29 @@
+
+import scala.quoted._
+import dotty.tools.dotc.quoted.Toolbox._
+
+object Test {
+
+  def main(args: Array[String]): Unit = {
+
+    def test[T](clazz: java.lang.Class[T]): Unit = {
+      val lclazz = clazz.toExpr
```
Comment (biboudis) [inline tests/run-with-compiler/i3947b.scala]: Same.

File: tests/run-with-compiler/i3947.scala (hunk 4 above)
Comment (nicolasstucki) [reply tests/run-with-compiler/i3947.scala]: Oh, I see. I lifted it directly in the quote. I will remove it or use it instead.

Comment (nicolasstucki) [reply tests/run-with-compiler/i3947.scala]: Done

File: tests/run-with-compiler/i3947b.scala (hunk 5 above)
Comment (nicolasstucki) [reply tests/run-with-compiler/i3947b.scala]: Done

**Comment by nicolasstucki** [comment]
Rebased

**Pull request merged**